### ISSUE TITLE
Issues with Patch commands

### ISSUE THREAD

**Issue by MMJM** [FusionAuth/fusionauth-netcore-client#37]
I think there is an issue with the Patch commands. In the client for example in the patchRegistration method it is expecting an IDictionary<string, object> for the request parameter but the Rest API according to the documentation is expecting a JSON which seems to be the RegistrationRequest type.

From what I can see the issue comes from the Client Builder in the _macros.ftl file. In that file is this code 

`      [#elseif language == "csharp"]
        [#local convertedType = convertType(param.javaType, language)/]
          [#if api.method == "patch" && param.javaType?ends_with("Request")]
            [#local convertedType = "IDictionary<string, object>"/]
          [/#if]
          [#local result = result + [convertedType + " " + param.name]/]
`

This changes what appears to be the correct type in the builder json files to an IDictonary. I think this may need to be removed.

**Comment by robotdan** [comment]
The reason for using `IDictionary<string, object>` is to allow you to build key value pairs that you want to use to build your PATCH request. 

For example, if you want to change only the name of an application using a `PATCH` - you would want to ensure that the JSON that is sent across the wire looks like this:

```
{
  "application": {
     "name": "changed"
   }
}
```

If you were to build up the `ApplicationRequest.cs` object and then serialize it - it will contain empty arrays, or other default values. This will modify the request and it will not behave how you want. 

By requiring you to enter key value pairs, we allow the JSON serialization to work correctly. This is essentially a limitation of the current implementation in .NET core and FusionAuth PATCH support. 

To do this correctly with typed objects, we'd have to ensure the .NET domain object has no default values, and then instruct the serializer to omit empty objects, empty arrays etc from the resulting JSON to ensure that the PATCH is performed correctly. 

There are a bunch of open issue with `PATCH` support in the main `fusionauth-issues` repo that we hope to solve at some point.  Once we have support for RFC 7396 ( https://github.com/FusionAuth/fusionauth-issues/issues/441 ) this may open up some additional options for configuring JSON serializer to allow us to use typed domain objects for `PATCH`. 

If you want to use typed objects, you can optionally perform a retrieve and then an update (`GET` and then `PUT`) instead of a `PATCH`.

**Comment by mooreds** [comment]
This has happened often enough that I added some doco to the relevant client libs paraphrasing what @robotdan wrote: https://github.com/FusionAuth/fusionauth-site/pull/458

**Comment by MMJM** [comment]
@robotdan @mooreds I get that was the intention but when I used it as intended by providing say "name" as the key and "changed" as the value it just fails.

**Comment by mooreds** [comment]
Hmmm. We have an example application which shows patch behavior: https://github.com/FusionAuth/fusionauth-example-netcore

From that application:

```
        static ClientResponse<UserResponse> patch(FusionAuthSyncClient client, String newfavColor, User userToPatch)
        {
	    Dictionary<string, object> request = new Dictionary<string, object>();
	    Dictionary<string, object> user = new Dictionary<string, object>();
	    Dictionary<string, object> data = new Dictionary<string, object>();
            user.Add("data", data);
	    data.Add("favoriteColor", newfavColor); 
            data.Add("recoveryCode", "recoveryCode");
            user.Add("firstName", "D");
            request.Add("user", user);
	    string json = JsonConvert.SerializeObject(request);
            Console.WriteLine(json);

            return client.PatchUser(userToPatch.id, request);
        }
```

I'm not an expert at .NET, so maybe that isn't idiomatic, but does that work for you?

**Comment by MMJM** [comment]
@mooreds Ah that is your usage. That is basically the same as what I had you are nesting dictionaries to get the JSON structure. I was under the assumption that I could just provide a key value pair for the item I wanted to change like ("favoriteColor", newfavColor) but from your example and mine that I have now tested it needs to be in sub dictionaries so that the dictionary containing "favoriteColor" needs to be added to a dictionary for "user". That makes sense but wasn't so clear.

I had also tried "user.favoriteColor" as a key.

Anyway it works using Dictionaries if you follow the need for sub dictionaries. Maybe mention that in the docs.

Thanks for the help and quick replies
